controllers/helpers.go

```diff
@@ -121,3 +129,215 @@ func GetObjectsToRequestsByNamespaceAndClusterName(ctx context.Context, c client
 	}
 	return results
 }
+
+// Returns true if a and b point to the same object
+func referSameObject(a, b metav1.OwnerReference) bool {
+	aGV, err := schema.ParseGroupVersion(a.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	bGV, err := schema.ParseGroupVersion(b.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	return aGV.Group == bGV.Group && a.Kind == b.Kind && a.Name == b.Name
+}
+
+// GetCloudProviderSecret returns the required azure json secret for the provided parameters. It differentiates between
+func GetCloudProviderSecret(d azure.ClusterDescriber, namespace, name string, identityType infrav1.VMIdentity, userIdentityID string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      fmt.Sprintf("%s-azure-json", name),
+			Labels: map[string]string{
+				string(infrav1.ResourceLifecycleOwned): "",
+			},
+		},
+	}
+
+	var data []byte
+	var err error
+
+	switch identityType {
+	case infrav1.VMIdentitySystemAssigned:
+		data, err = systemAssignedIdentityCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityUserAssigned:
+		if len(userIdentityID) < 1 {
+			return nil, errors.New("expected a non-empty userIdentityID")
+		}
+		data, err = userAssignedIdentityCloudProviderConfig(d, userIdentityID)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityNone:
+		data, err = servicePrincipalCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	secret.Data = map[string][]byte{
+		"azure.json": data,
+	}
+
+	return secret, nil
+}
+
+func servicePrincipalCloudProviderConfig(d azure.ClusterDescriber) ([]byte, error) {
+	return json.MarshalIndent(newCloudProviderConfig(d), "", "    ")
+}
+
+func systemAssignedIdentityCloudProviderConfig(d azure.ClusterDescriber) ([]byte, error) {
+	config := newCloudProviderConfig(d)
+	config.AadClientID = ""
+	config.AadClientSecret = ""
+	config.UseManagedIdentityExtension = true
+	return json.MarshalIndent(config, "", "    ")
+}
+
+func userAssignedIdentityCloudProviderConfig(d azure.ClusterDescriber, identityID string) ([]byte, error) {
+	config := newCloudProviderConfig(d)
+	config.AadClientID = ""
+	config.AadClientSecret = ""
+	config.UseManagedIdentityExtension = true
+	config.UserAssignedIdentityID = identityID
+	return json.MarshalIndent(config, "", "    ")
+}
+
+func newCloudProviderConfig(d azure.ClusterDescriber) *CloudProviderConfig {
+	return &CloudProviderConfig{
+		Cloud:                        d.CloudEnvironment(),
+		AadClientID:                  d.ClientID(),
+		AadClientSecret:              d.ClientSecret(),
```

Comment:
should we consider not putting the credentials on non-control plane nodes if they aren't used? based on https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/773#issuecomment-659082883

Reply:
I think we can remove almost all of the values on worker nodes, but I hadn't tested that yet.

Reply:
want me to try and tackle in this PR, or ok to slim down in follow up? I think it won't be too difficult, but I haven't double-checked the kubelet code for any edge cases like we hit with VMAS/VMSS/no zones

Reply:
let's tackle in a follow up since the behavior in this PR is equivalent to current state of things